File: vantage-backend/sdk/vantage/utils/queue.py
"""
Async event queue — batches events and flushes to ingest server.
Zero latency impact on the application.
"""
from __future__ import annotations
import json, logging, queue, threading, time, urllib.error, urllib.request
from vantage.models.event import VantageEvent

logger = logging.getLogger("vantage.queue")
SDK_VERSION = "1.0.0"


class EventQueue:
    def __init__(self, api_key: str, ingest_url: str, flush_interval: float = 2.0,
                 batch_size: int = 50, debug: bool = False):
        self.api_key        = api_key
        self.ingest_url     = ingest_url.rstrip("/")
        self.flush_interval = flush_interval
        self.batch_size     = batch_size
        self.debug          = debug
        self._q: queue.Queue[VantageEvent] = queue.Queue(maxsize=10_000)
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, daemon=True, name="vantage-flush")
        self._thread.start()

    def enqueue(self, event: VantageEvent) -> None:
        try:
            self._q.put_nowait(event)
            qsize = self._q.qsize()
            if self.debug:
                logger.debug("[vantage] %s | %s | %d tok | $%.6f",
                             event.provider, event.model,
                             event.usage.total_tokens, event.cost.total_cost_usd)
            # Warn at 80% capacity before silent drops occur
            if qsize >= 8_000 and qsize % 500 == 0:
                logger.warning(
                    "Vantage queue at %d/10000 — events will be dropped at capacity. "
                    "Consider reducing flush_interval or batch_size.",
                    qsize,
                )
        except queue.Full:
            logger.warning("Vantage queue full (10000) — event dropped")

    def flush_sync(self) -> None:
        batch: list[VantageEvent] = []
        try:
            while len(batch) < self.batch_size:
                batch.append(self._q.get_nowait())
        except queue.Empty:
            pass
        if batch:
            self._send(batch)

    def _run(self) -> None:
        while True:
            time.sleep(self.flush_interval)
            try:
                self.flush_sync()
            except Exception as e:
                logger.warning("Flush error: %s", e)

    def _send(self, events: list[VantageEvent]) -> None:
        payload = json.dumps({
            "events": [e.to_dict() for e in events],
            "sdk_version": SDK_VERSION,
        }).encode()
        req = urllib.request.Request(
            f"{self.ingest_url}/v1/events", data=payload, method="POST",
            headers={"Content-Type": "application/json",
                     "Authorization": f"Bearer {self.api_key}"},
        )
        try:
            with urllib.request.urlopen(req, timeout=5) as r:
                if self.debug:
                    logger.debug("Flushed %d events → %s", len(events), r.status)
        except Exception as e:
            logger.warning("Ingest failed: %s", e)

File: vantage-js-sdk/src/utils/queue.ts
import { VantageEvent, flattenEvent } from "../models/event.js";

const SDK_VERSION = "1.0.0";
const MAX_RETRIES = 3;

interface QueuedEvent {
  event: VantageEvent;
  retries: number;
}

export class EventQueue {
  private queue: QueuedEvent[] = [];
  private readonly maxSize = 10_000;
  private timer: ReturnType<typeof setInterval> | null = null;
  private _flushing = false;

  constructor(
    private readonly apiKey: string,
    private readonly ingestUrl: string,
    private readonly flushInterval = 2000,
    private readonly batchSize = 50,
    private readonly debug = false
  ) {}

  start(): void {
    if (this.timer) return;
    this.timer = setInterval(() => this.flush(), this.flushInterval);
    // Flush on process exit
    if (typeof process !== "undefined") {
      process.on("beforeExit", () => this.flush());
    }
  }

  stop(): void {
    if (this.timer) { clearInterval(this.timer); this.timer = null; }
  }

  enqueue(event: VantageEvent): void {
    if (this.queue.length >= this.maxSize) {
      if (this.queue.length === this.maxSize) {
        console.warn("[vantage] Event queue full (10k). Dropping event.");
      }
      return;
    }
    if (this.queue.length >= this.maxSize * 0.8 && this.queue.length % 500 === 0) {
      console.warn(`[vantage] Queue at ${this.queue.length} events — consider flushing more frequently.`);
    }
    this.queue.push({ event, retries: 0 });
    if (this.debug) console.debug(`[vantage] Enqueued event ${event.eventId} (queue size: ${this.queue.length})`);
  }

  flush(): void {
    if (this._flushing) return;
    if (this.queue.length === 0) return;
    this._flushing = true;
    const batch = this.queue.splice(0, this.batchSize);
    this._send(batch)
      .catch((err) => {
        if (this.debug) console.warn("[vantage] Flush error:", err);
        // Re-queue only items that have not exceeded max retries
        const retryable = batch
          .map((item) => ({ ...item, retries: item.retries + 1 }))
          .filter((item) => {
            if (item.retries >= MAX_RETRIES) {
              if (this.debug) console.warn(`[vantage] Dropping event ${item.event.eventId} after ${MAX_RETRIES} failures.`);
              return false;
            }
            return true;
          });
        this.queue.unshift(...retryable);
      })
      .finally(() => { this._flushing = false; });
  }

  private async _send(items: QueuedEvent[]): Promise<void> {
    const events = items.map((i) => i.event);
    const body = JSON.stringify({
      events: events.map(flattenEvent),
      sdk_version: SDK_VERSION,
      sdk_language: "typescript",
    });

    try {
      const res = await fetch(`${this.ingestUrl}/v1/events`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.apiKey}`,
        },
        body,
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      if (this.debug) console.debug(`[vantage] Sent ${items.length} events → ${res.status}`);
    } catch (err) {
      if (this.debug) console.warn("[vantage] Ingest failed:", err);
      throw err; // re-throw so flush() re-queues the batch
    }
  }
}
